gcp-pubsub/src/main/java/io/micronaut/gcp/pubsub/intercept/PubSubConsumerAdvice.java

```diff
@@ -0,0 +1,181 @@
+
+        if (subscriptionAnnotation != null) {
+            String subscriptionName = subscriptionAnnotation.getRequiredValue(String.class);
+            ProjectSubscriptionName projectSubscriptionName = PubSubSubscriptionUtils.toProjectSubscriptionName(subscriptionName, googleCloudConfiguration.getProjectId());
+            String defaultContentType = subscriptionAnnotation.get("contentType", String.class).orElse("");
+            MessageReceiver receiver = (PubsubMessage message, AckReplyConsumer consumer) -> {
+                String messageContentType = message.getAttributesMap().getOrDefault("Content-Type", "");
+                String contentType = Optional.of(messageContentType).orElse(defaultContentType);
+                DefaultPubSubAcknowledgement pubSubAcknowledgement = new DefaultPubSubAcknowledgement(consumer);
+
+                PubSubConsumerState consumerState = new PubSubConsumerState(message, consumer,
+                        projectSubscriptionName, contentType);
+                try {
+                    BoundExecutable executable = null;
+                    try {
+                        executable = binder.bind(method, binderRegistry, consumerState);
+                    } catch (Exception ex) {
+                        handleException(new PubSubMessageReceiverException("Error binding message to the method", ex, bean, consumerState));
+                    }
+                    executable.invoke(bean); // Discard result
+                    if (!hasAckArg) { // if manual ack is not specified we auto ack message after method execution
+                        pubSubAcknowledgement.ack();
+                    } else {
+                        if (!pubSubAcknowledgement.isMessageAcked()) {
+                            logger.warn("Method {} was executed and no message acknowledge detected, message will be redelivered.", method.getName());
+                        }
+                    }
+                } catch (Exception e) {
+                    handleException(new PubSubMessageReceiverException("Error handling message", e, bean, consumerState));
+                }
+            };
+            try {
+                SubscriberInterface subscriber = this.subscriberFactory.createSubscriber(projectSubscriptionName, receiver);
+                registeredSubscribers.put(projectSubscriptionName.toString(), subscriber);
+            } catch (Exception e) {
+                throw new PubSubListenerException("Failed to create subscriber", e);
+            }
+
+        }
+
+    }
+
+    private void handleException(PubSubMessageReceiverException ex) {
+        Object bean = ex.getListener();
+        if (bean instanceof PubSubMessageReceiverExceptionHandler) {
+            ((PubSubMessageReceiverExceptionHandler) bean).handle(ex);
+        } else {
+            exceptionHandler.handle(ex);
+        }
+    }
+
+    @PreDestroy
+    @Override
+    public void close() throws Exception {
```

Comment:
The subscribers are still being stored in separate maps. I think it would be better for the close logic to happen within the subscriber factory. Alternatively you can move the "single subscriber" logic to this class and the factory wouldn't keep a reference to the subscriber after its created

Reply:
Thanks, moved it to SubscriberFactory where it indeed makes more sense.